### PR TITLE
Fix Channels not showing up when Message attachment has malformed data

### PR DESCRIPTION
**Pull request by DominikBucher12**
# What this PR do:
- Fixes issue where malformed attachment data could cause Channel not to appear
# How to test it
1. Open 2 simulator device at same time and on each run the same instance of Demo App
1. Select any user 
1. Exit one of the simulators (A)
1. From the other one (B) send some messages for the selected user account
1. Start A again with the app and log in to the same user
1. Observe that all channels are displayed for the same user on both devices

### CODE REVIEW

File: Sources/StreamChat/Database/DTOs/MessageDTO.swift [hunk 1]
```diff
@@ -324,11 +324,19 @@ extension NSManagedObjectContext: MessageDatabaseSession {
...
-                let dto = try saveAttachment(payload: attachment, id: id)
-                return dto
+                do {
+                    let dto = try saveAttachment(payload: attachment, id: id)
+                    return dto
+                } catch {
+                    log.assertationFailure(
+                        "Failed to save attachment: \(attachment). Please check saveAttachment function at " +
+                            "AttachmentDTO.swift"
+                    )
```
Comment (VojtaStavik) [inline Sources/StreamChat/Database/DTOs/MessageDTO.swift:337]: No need to crash here, please use just `log.error`

**Review by VojtaStavik — APPROVED** [review]
Please also add a test for the new functionality. Thanks

File: Sources/StreamChat/Database/DTOs/MessageDTO.swift (hunk 1 above)
Comment (DominikBucher12) [reply Sources/StreamChat/Database/DTOs/MessageDTO.swift]: Fixed.

**Review by nuno-vieira — APPROVED** [review]
LGTM

File: Sources/StreamChat/Utils/OptionalDecodable.swift [hunk 2]
```diff
@@ -0,0 +1,19 @@
+//
+// Copyright © 2021 Stream.io Inc. All rights reserved.
+//
+
+import Foundation
+
+struct OptionalDecodable<Base: Decodable>: Decodable {
```
Comment (VojtaStavik) [inline Sources/StreamChat/Utils/OptionalDecodable.swift:11]: Please add a comment explaining what is this and why/when to use it.

File: Sources/StreamChat/APIClient/Endpoints/Payloads/MessagePayloads.swift [hunk 3]
```diff
@@ -92,7 +92,7 @@ class MessagePayload<ExtraData: ExtraDataTypes>: Decodable {
         reactionScores = try container
             .decodeIfPresent([String: Int].self, forKey: .reactionScores)?
             .mapKeys { MessageReactionType(rawValue: $0) } ?? [:]
-        attachments = try container.decode([AttachmentPayload].self, forKey: .attachments)
+        attachments = try container.decode([OptionalDecodable<AttachmentPayload>].self, forKey: .attachments).compactMap(\.base)
```
Comment (VojtaStavik) [inline Sources/StreamChat/APIClient/Endpoints/Payloads/MessagePayloads.swift:95]: Please add a comment explaining what is happening here.

File: Sources/StreamChat/APIClient/Endpoints/Payloads/MessagePayloads_Tests.swift [hunk 4]
```diff
@@ -34,6 +35,33 @@ class MessagePayload_Tests: XCTestCase {
...
+        XCTAssertEqual(payload.mentionedUsers.map(\.id), [])
+        XCTAssertEqual(payload.threadParticipants.map(\.id), ["josh"])
+        XCTAssertEqual(payload.replyCount, 0)
+        XCTAssertEqual(payload.extraData, .defaultValue)
+        XCTAssertEqual(payload.latestReactions.count, 1)
+        XCTAssertEqual(payload.ownReactions.count, 1)
+        XCTAssertEqual(payload.reactionScores, ["love": 1])
+        XCTAssertEqual(payload.isSilent, true)
+        XCTAssertEqual(payload.channel?.cid.rawValue, "messaging:channel-ex7-63")
+        XCTAssertEqual(payload.quotedMessage?.id, "4C0CC2DA-8AB5-421F-808E-50DC7E40653D")
```
Comment (VojtaStavik) [inline Sources/StreamChat/APIClient/Endpoints/Payloads/MessagePayloads_Tests.swift:63]: The JSON contains 2 attachments and only one of them is broken. Should we assert the other one is serialized correctly?

**Review by VojtaStavik — COMMENTED** [review]
Just nits :shipit:

**Review by VojtaStavik — APPROVED** [review]

**Review by b-onc — APPROVED** [review]
I like the solution, pity that this isn't JSONDecoder's default behavior....

File: Sources/StreamChat/APIClient/Endpoints/Payloads/MessagePayloads.swift (hunk 3 above)
Comment (DominikBucher12) [reply Sources/StreamChat/APIClient/Endpoints/Payloads/MessagePayloads.swift]: Fixed.

File: Sources/StreamChat/Utils/OptionalDecodable.swift (hunk 2 above)
Comment (DominikBucher12) [reply Sources/StreamChat/Utils/OptionalDecodable.swift:15]: Fixed.

File: Sources/StreamChat/APIClient/Endpoints/Payloads/MessagePayloads_Tests.swift (hunk 4 above)
Comment (DominikBucher12) [reply Sources/StreamChat/APIClient/Endpoints/Payloads/MessagePayloads_Tests.swift:63]: Checking only for attachment count if there is just one correct attachment...

**Pull request merged**